compiler/src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -212,6 +212,8 @@ trait TypeOps { this: Context => // TODO: Make standalone object.
               tp1.derivedAppliedType(
                 mergeRefinedOrApplied(tycon1, tycon2),
                 ctx.typeComparer.lubArgs(args1, args2, tycon1.typeParams))
+            case AndType(tp21, tp22) =>
```

Comment:
This handles having tp1 be an AppliedType and tp2 be an AndType, but it seems that we'd need similar special cases when tp1 is a RefinedType and tp2 is an AndType, etc. Instead maybe all calls to `fail` in this method could be replaced by a fallback that handles the recursion when tp2 is an AndType.